File: Android/app/src/main/java/com/dreamproject/inwords/data/interactor/authorisation/AuthorisationWebInteractor.java
package com.dreamproject.inwords.data.interactor.authorisation;

import com.dreamproject.inwords.data.entity.UserCredentials;
import com.dreamproject.inwords.data.source.WebService.AuthenticationError;
import com.dreamproject.inwords.data.source.WebService.TokenResponse;
import com.dreamproject.inwords.data.source.WebService.WebRequests;
import com.dreamproject.inwords.util.ErrorBodyFormatter;

import java.net.UnknownHostException;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.Single;
import retrofit2.HttpException;

public class AuthorisationWebInteractor implements AuthorisationInteractor {
    private WebRequests webRequests;

    @Inject
    AuthorisationWebInteractor(WebRequests webRequests) {
        this.webRequests = webRequests;
    }

    @Override
    public Completable signIn(UserCredentials userCredentials) {
        return webRequests.setCredentials(userCredentials)
                .andThen(checkAuthToken(webRequests.getToken()));
    }

    @Override
    public Completable signUp(UserCredentials userCredentials) {
        return checkAuthToken(webRequests.registerUser(userCredentials));
    }

    private Completable checkAuthToken(Single<TokenResponse> authTokenSingle) {
        return authTokenSingle
                .onErrorResumeNext(e -> {
                    e.printStackTrace();

                    Throwable t;
                    if (e instanceof HttpException)
                        t = new AuthenticationError(ErrorBodyFormatter.getErrorMessage((HttpException) e));
                    else if (e instanceof UnknownHostException)
                        t = new RuntimeException("Network troubles");
                    else
                        t = new RuntimeException(e.getMessage());

                    return Single.error(t);

                })
                .flatMapCompletable(tokenResponse -> {
                    if (tokenResponse.isValid())
                        return CompletableObserver::onComplete;

                    return Completable.error(new RuntimeException("unhandled")); //TODO think
                });
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/domain/interactor/authorisation/AuthorisationWebInteractor.kt
package ru.inwords.inwords.domain.interactor.authorisation

import io.reactivex.Completable
import io.reactivex.Single
import retrofit2.HttpException
import ru.inwords.inwords.data.dto.UserCredentials
import ru.inwords.inwords.data.source.webService.AuthenticationException
import ru.inwords.inwords.data.source.webService.WebRequestsManager
import ru.inwords.inwords.data.source.webService.session.TokenResponse
import ru.inwords.inwords.domain.IntegrationInteractor
import ru.inwords.inwords.domain.util.getErrorMessage
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.inject.Inject

class AuthorisationWebInteractor @Inject
internal constructor(private val webRequestsManager: WebRequestsManager,
                     private val integrationInteractor: IntegrationInteractor) : AuthorisationInteractor {

    override fun signIn(userCredentials: UserCredentials): Completable {
        return webRequestsManager.getToken(userCredentials).interceptError().checkAuthToken()
    }

    override fun signUp(userCredentials: UserCredentials): Completable {
        return webRequestsManager.registerUser(userCredentials).interceptError().checkAuthToken()
    }

    private fun Single<TokenResponse>.interceptError(): Single<TokenResponse> {
        return onErrorResumeNext { e ->
            e.printStackTrace()

            val t = when (e) {
                is HttpException -> AuthenticationException(getErrorMessage(e), e.code())
                is UnknownHostException, is SocketTimeoutException -> RuntimeException("Network troubles")
                else -> RuntimeException(e.message)
            }

            Single.error(t)
        }
    }

    private fun Single<TokenResponse>.checkAuthToken(): Completable {
        return flatMapCompletable { tokenResponse ->
            if (tokenResponse.isValid) {
                integrationInteractor.getOnAuthCallback()
            } else {
                Completable.error(RuntimeException("unhandled")) //TODO think
            }
        }
    }
}
